### ISSUE TITLE
NVDA does not seem to read past end of page.

### ISSUE THREAD

**Issue by a-singer** [blindpandas/bookworm#3]
## Describe the Problem

	NVDA does not continue to read past the end of the current page, no matter which setting is chosen in the reading section of preferences. I understand that the setting says "when pressing play", and therefore is probably only applicable to the program's own TTS.
	There is no setting I can find which allows the entire page or book to be displayed for NVDA to read with the navigation functions of BW available. One must start the reading and, when it is done, press enter to turn the page and then start reading again. Would it be possible for the screen reader to read the entire text without interuption?

## To Reproduce

1. Open an Epub book with multiple chapters/pages.o 
2. SGo to one of the chapters and start reading with your screen reader, not with the program TTS.
3. The screen reader will not go on reading when the end of the page is reached.

## Expected behavior

I wish the screen reader to keep reading until told to stop.

## The File You've Opened

Epub

## Desktop (please complete the following information):

 - OS: Windows 7 64 Bit

 - Bookworm version 0.1.0b1

**Comment by mush42** [comment]
Hello,
I understand how having such feature is important for some users.
Unfortunately, implementing this is out of reach for now.
Nevertheless, I'll put it in my list, and try to find a way to capture the caret move events and act accordingly.

Thanks

**Comment by a-singer** [comment]
Thanks for the quick reply and the attention. I don't know how helpful this is, but Q-Read
http://www.getaccessibleapps.com
does this quite well. I say I don't know how helpful this is because the program seems to have been abandoned but it does what this issue suggests, and it may be worth reaching out to the author or examining the program if this is something you are willing to do and if the author would be willing to answer. Thanks again for the rapid and attentive reply.

**Comment by Simon818** [comment]
The majority of people are going to want to use their own screen-reader until more voice options become available on Windows. I like the built-in read aloud functions, but the only reason I'm able to use them effectively is because I used a registry hack to get the mobile microsoft voices running, and I bought the Code Factory version of Eloquence. Sounds anecdotal, but I say this because both options are out of reach for many users and unknown to many more. And the end result is that screen-readers are the clear winner for those people, so continuous reading is really important. So far, I like everything else about this app and this is the one thing holding me back from recommending it to other people I know.

**Comment by mush42** [comment]
Another priority feature.
We're trying not to let you down, but technical limitations are a bit hard to overcome. Nevertheless we're actively working to resolve this particular issue.

**Comment by Simon818** [comment]
Thanks for this comment. I didn't get the impression it was a priority at first, which is why I said this. I definitely see what you mean about the technical limitations; there's basically no way to make it work while still only rendering a page at a time on-screen.

What are the drawbacks of rendering all the text at once and simply controlling where the cursor lands in order to "fake" the current navigation we have now? This is the route QRead decided to go, and even with a 3.5 million word epub, it doesn't seem to slow down at all.

**Comment by mush42** [comment]
I intentionally avoided using QRead-style rendering. "Not slow" is not the same thing as "fast".
I didn't test this, but I can guarantee you that the resource usage of QRead is at least an order of magnitude higher than Bookworm's resource usage.
Moreover, QRead is significantly slower in a test machine that I use for benchmarks. That machine runs Windows 10 Pro 64-bit in an Intel core-i2-duo processor.

**Comment by Simon818** [comment]
So for fun, I loaded a 3.5 million word epub file into QRead and Bookworm. Bookworm loaded the file very quickly (less than ten seconds), but the memory usage skyrocketed to almost 500 mb. QRead took much longer to load it, but the memory was about 87 mb. That's my very unscientific test, in case you were curious.

**Comment by mush42** [comment]
What about having multiple reading modes?

We can keep the current, pagination-based, reading mode, and add a setting to switch to a, continuous, reading mode, the same way Adobe Acrobat does for visual rendering.

This requires a lot of refactoring. But I think it would be worthwhile.

**Comment by Simon818** [comment]
As long as you think that's possible, I don't see any drawbacks. It would let us compare the performance between the two modes as well, which might be useful.

**Comment by a-singer** [comment]
Hi,


    I really like this idea. It is also worth noting that the idea of "page" is a bit fluid in Epub files. I should also like to congratulate you for thinking of people with lower spec machines, I threw my hands up and gave up on Kurzweil 1000 because, even on my machine, which is fairly high-spec, it took forever to open some larger epubs. QRead openned them without issues in a much shorter period of time. The same may be said of getting a table of contents, etc, QR was much much faster than K1000 and BW is even faster. Having said that, if there is a way to have continuous reading, I would, though obviously just as one user, take the performance hit.

Thanks again,


On 9/17/2019 4:46 AM, Musharraf wrote:

What about having multiple reading modes?

We can keep the current, pagination-based, reading mode, and add a setting to switch to a, continuous, reading mode, the same way Adobe Acrobat does for visual rendering.

This requires a lot of refactoring. But I think it would be worthwhile.

**Comment by AntonioCezar** [comment]
Hello. I do not know if it's the same thing, but I tried not render pages with control + R one very large file here with Bookworm, and the application has taken a crash. Perhaps a stopgap possibility was to get NVDA automatically continue reading after it gives enter or backspace to change the page. I've thought about the possibility of including creating an add-on for this purpose, but I do not know to program in Pyton, in which case, I would have to convince someone who can work with add-ons to implement this feature ... So why in tests I've done, is using the voice of Cortana, either using a test version of Eloquence of the Code factory with the internal mechanism talking Bookworm, I can not have such a fluent reading as when using continuous reading NVDA. At that point, the QReader can be quite good, and I just do not make use of QReader in my day to day, because it divides the text Numas imaginary line breaks, which greatly truncates reading and also disturbs me much time to colpiar any piece of text for a citation academic work, for example, which means that I have always preferred to convert PDF to txt with pdftotext, which always gives me the lines almost without any imaginary break by default ... Thanks for your work with the Bookworm.

**Comment by mush42** [comment]
Hi everyone,

Good news regarding this issue. Based on the suggestion over at wxWidgets/Phoenix#1420, this issue has been resolved.

Continuous reading mode has been added. When this mode is in use, the following behavior should be expected:

* When the screen reader's say-all mode is active, pages are turned automatically for you. Moreover, turning pages will not interrupt the speech.
* Pressing the up/down arrow moves to previous/next page respectively.

**Comment by AntonioCezar** [comment]
Hi. I do not find the ondde link i can download this trial version with automatic scrolling implemented page. But right now, I would like to make another suggestion, so that the navigation commands in the Bookworm are closer to those commands that I usually use in all other applications on Windows. For example, I use the Ende to go to the end of a line; Home to go to the beginning of a line; the up or down to go to the previous line or higher; the control over page-up to go to the previous page; the control over page-down, to go to the next page; Control over Home to go to Nicio the document; control over ende to go to the end of the document. I ask this change, because I often use the commands with which I'm used to reading in any application on Windows, but when I'm at the Bookworm, the results are totally different from the standard Windows ... Thanks.

**Comment by mush42** [comment]
Hi @a-singer, @Simon818, and @AntonioCezar:

Here is a try build that contains the latest changes. This is a portable copy of Bookworm, you can try it without installing.
Please note that this is alpha quality code.

[Bookworm-0.1b3-portable.zip](https://github.com/mush42/bookworm/files/3803947/Bookworm-0.1b3-portable.zip)

Regards,

**Comment by AntonioCezar** [comment]
Hi. I still can not use the control over ende to go to the end of the document nor the control over home to go to the beginning of the document. I also can not use the page-up to go to the previous page, but I already can use the control over page-down to go to the next page. The enter and backspace also serve to navigate through the pages in the same way it was in the previous version ... Thank you for the work.

**Comment by AntonioCezar** [comment]
Hi, again. I realized that if I use Alt plus home or ende, I sail quickly to the beginning or end of the document. However, I still can not return to the previous page with control over page-up, although I have the possibility of returning to the previous page with the backspace, which continues to be a reasonable solution because, unlike home , ende and arrows up / down / left / right, which are important to navigate the text in a PDF in read-only mode, enter and backspace do not serve me for nothing ... Thanks.

**Comment by AntonioCezar** [comment]
Sorry, but my goal is to help improve the application. So I bring yet another issue for discussion: In some texts, I need to do a more enhanced reading, preferably paragraph by paragraph, for study purposes. So when I use the command control more arrow down or control more arrow-up, to navigate by paragraphs, I can not get the desired effect, because the Bookworm puts a nonexistent line breaks in the text. I know they are nonexistent line breaks, because if I convert the same material using another application, these nonexistent line breaks do not appear. I do not know if you have any technical limitation to solve this, but if there was a way to suppress these breaks nonexistent lines, Bookworm would be much better ... Thank you.

**Comment by mush42** [comment]
Hi @AntonioCezar 

Please follow up on issue #24 as this issue will be closed soon.

Regards,

**Comment by AntonioCezar** [comment]
Hi. I have another problem with the automatic scrolling of pages. Some books I'm reading in Bookworm, he jumps / skips pages without reading their contents, as if nothing in these pages, but when I'm reading the same book in other PDF readers, these pages are read normally without any problems. I know it will not help much, because of the language difference, because the books I read are in Portuguese of Brazil. Still, I'll put down a link to a book that can be downloaded from the internet, so you can reproduce the problem there in your system. Note that in that book, put on continuous reading with NVDA from page 20, for example, he will skip two pages, will read a whole page, and will jump another three pages, to re-read again, but it happens in various parts of the book. With my anticipated thanks, it is below the book PDF link, as mentioned above, for the case to be interesting you take a look at this:
https://forumdeconcursos.com/wp-content/uploads/wpforo/attachments/2/1755-PIERLUIGI-PIAZZI-VOL-4-INTELIGNCIA-EM-CONCURSOS.pdf

**Comment by a-singer** [comment]
Hi,


Thanks very much. I have found a book in which the pages are not properly turned for some reason during say-all. Is there a method by which I can report exactly what is happening and provide the book?

Thanks.



On 11/4/2019 7:23 AM, Musharraf wrote:

Hi @a-singer<https://github.com/a-singer>, @Simon818<https://github.com/Simon818>, and @AntonioCezar<https://github.com/AntonioCezar>:

Here is a try build that contains the latest changes. This is a portable copy of Bookworm, you can try it without installing.
Please note that this is alpha quality code.

Bookworm-0.1b3-portable.zip<https://github.com/mush42/bookworm/files/3803947/Bookworm-0.1b3-portable.zip>

Regards,

**Comment by mush42** [comment]
Hi @a-singer,

I am aware of this problem. Currently working on fixing it.

**Comment by mush42** [comment]
Hi everyone,

Sorry for not updating you on this issue. Some work-related matters, as usual!

I fixed the bug of page skipping, and I'm using the continuous reading feature happily for two weeks without any issues.

We're working on releasing a new version soon, but in the mean time here is a copy with the latest changes.


[Bookworm-07122019-portable.zip](https://github.com/mush42/bookworm/files/3935120/Bookworm-07122019-portable.zip)

Regards,

**Comment by a-singer** [comment]
Hi,


Let me first reiterate how grateful I am that you take the time to work on this project. It is an excellent reader. I am, unfortunately, still having the issue with reading just stopping. Is this supposed to be fixed in the new build? I may say that, and I don't know if this is my imagination, this problem seems to be reduced in frequency in the new build. Should it be resolved?


Thanks,

Aman


On 12/7/2019 8:48 AM, Musharraf wrote:

Hi everyone,

Sorry for not updating you on this issue. Some work-related matters, as usual!

I fixed the bug of page skipping, and I'm using the continuous reading feature happily for two weeks without any issues.

We're working on releasing a new version soon, but in the mean time here is a copy with the latest changes.

Bookworm-07122019-portable.zip<https://github.com/mush42/bookworm/files/3935120/Bookworm-07122019-portable.zip>

Regards,

**Comment by AntonioCezar** [comment]
Hello. Here, I not only continue to have problems stops at the end of a 
few pages, but also I realized that he jumped at least one page of a few 
that I read that same book in Portuguese that I sent to exemplify the 
Bookworm behavior ... Thank you.


------------------------------------------------------------------------

Por favor, assine nossa petição no link abaixo, compartilhe com todos os 
seus amigos, e colabore com a gente! Em favor do suporte para teclados 
Braille com digitação multitoque diretamente na tela de dispositivos 
Android para qualquer leitor de telas, principalmente para o Talkback!

https://secure.avaaz.org/po/petition/Google_Inc_Talkback_Development_Team_Google_Disability_Support_Team_PRECISAMOS_DO_BRAILLE_NO_ANDROID_TAMBEM/

Em 09/12/2019 18:55, a-singer escreveu:

**Comment by mush42** [comment]
Hello all,

The feature works perfectly in my test machines. To diagnose this issue, could you give me more information about the following:

1. Your OS version and bitness, and Bookworm and its bitness
2. The screen reader you're using
3. What speech synthesizer you're using with your screen reader
4. Does the stop occur when the reading position reaches a blank page? or Bookworm stops randomly at any position.
5. If you're using NVDA: which of the following methods do you use to activate say-all?
(a). Numpad Plus
(b). NVDA + A or NVDA + Down Arrow


Regards,

**Comment by AntonioCezar** [comment]
Here are my answers to your questions:

Em 10/12/2019 09:42, Musharraf escreveu:

**Comment by a-singer** [comment]
Thanks for looking into this, here are my answers.


1. Your OS version and bitness, and Bookworm and its bitness


Answer: Windows 7 X64, the exact build of BookWorm last sent on this issue, having file name

Bookworm-07122019-portable.zip


2. The screen reader you're using


Answer: NVDA 2019.2.1 with Eloquence and Espeak, Jaws 2019 with Eloquence. Happens with all screen readers and synthesizers.


3. What speech synthesizer you're using with your screen reader


Answer: Eloquence and ESpeak, stops at the same point with both.


4. Does the stop occur when the reading position reaches a blank page? or Bookworm stops randomly at any position.


Answer: So far, the stop has occurred at the ends of pages, not at a blank page. I have not seen reading stop in the middle of a page. In one of my books I can make it happen on demand at the end of a certain page.


5. If you're using NVDA: which of the following methods do you use to activate say-all?
(a). Numpad Plus
(b). NVDA + A or NVDA + Down Arrow


Answer: B, NVDA+A

I hope that's of use.


On 12/10/2019 7:42 AM, Musharraf wrote:

Hello all,

The feature works perfectly in my test machines. To diagnose this issue, could you give me more information about the following:

  1.  Your OS version and bitness, and Bookworm and its bitness
  2.  The screen reader you're using
  3.  What speech synthesizer you're using with your screen reader
  4.  Does the stop occur when the reading position reaches a blank page? or Bookworm stops randomly at any position.
  5.  If you're using NVDA: which of the following methods do you use to activate say-all?
(a). Numpad Plus
(b). NVDA + A or NVDA + Down Arrow

Regards,

**Comment by mush42** [comment]
Hello all,

This issue has been closed automatically as a part of v0.1b4 release cycle. However I'm reopening it again. It will not be marked as fixed until all users are satisfied.
